### PR TITLE
feat(addon-doc): ability to copy custom API configuration code

### LINKED ISSUE #5475: Sponsored issue: Ability to copy custom API configuration code
A way to directly copy the code implementation when testing out component custom API configuration on the site.

## Priority Support

- @farbaks is using  [Mintycode](https://mintycode.io) to fund this issue.
- If you would like to accept amount bounty for solving this issue join [Mintycode](https://mintycode.io/profile?action=CREATOR_SUPPORT_MODAL&requestId=2d45fe77-a900-455e-bc12-4efd9b93e263&owner=taiga-family&name=taiga-ui).
- Thank you in advance for helping.

[mintycode](https://mintycode.io)

### PR DESCRIPTION
**Pull request by MillerSvt**
## PR Type

What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature
- [ ] Refactoring
- [ ] Build or CI related changes
- [ ] Tests related changes
- [ ] Documentation content changes

## What is the current behaviour?
no way to copy custom API configuration code

Closes #5475 

## What is the new behaviour?
ability to copy custom API configuration code
image

---
Reopening: #5816

### CODE REVIEW

**Comment by splincode** [comment]
@MillerSvt redesign please

image


to

image

**Comment by splincode** [comment]
image

change to

```
[markerHandler]="fn"
```

**Comment by MillerSvt** [comment]
@splincode Should we replace string inputs `[foo]="'bar'"` to `foo="bar"`?

**Comment by MillerSvt** [comment]
> @MillerSvt redesign please

@splincode There is a small problem: If we remove border-radius from resizable frame, then it will look incorrect in sticky mode.

Now it looks like this:
image

If we remove border-radius it will be look like this:
image

I can't add `overflow: hidden`, because of break sticky positioning.

File: projects/addon-doc/components/documentation/directives/documentation-property-connector.directive.ts [hunk 1]
```diff
@@ -93,14 +102,33 @@ export class TuiDocDocumentationPropertyConnectorDirective<T>
...
+        }
+
+        this.#updateApiHostProperty();
     }
 
     onValueChange(value: T): void {
         this.documentationPropertyValue = value;
         this.documentationPropertyValueChange.emit(value);
         this.setQueryParam(value);
+        this.#updateApiHostProperty();
```
Comment (splincode) [inline projects/addon-doc/components/documentation/directives/documentation-property-connector.directive.ts:131]: please don't use `#` right now, we don't use new private fields yet

File: projects/addon-doc/components/documentation/directives/documentation-property-connector.directive.ts [hunk 2]
```diff
@@ -139,15 +167,68 @@ export class TuiDocDocumentationPropertyConnectorDirective<T>
...
 
-        tree.queryParams = {
-            ...tree.queryParams,
-            [propName]: computedValue,
-        };
+        if (computedDefaultValue === computedValue) {
+            delete tree.queryParams[propName];
+        } else {
+            tree.queryParams[propName] = computedValue;
+        }
```
Comment (splincode) [inline projects/addon-doc/components/documentation/directives/documentation-property-connector.directive.ts:186]: why not immutable structure?

File: projects/addon-doc/components/documentation/directives/documentation-property-connector.directive.ts [hunk 3]
```diff
@@ -139,15 +167,68 @@ export class TuiDocDocumentationPropertyConnectorDirective<T>
...
 
         this.locationRef.go(this.urlStateHandler(tree));
     }
+
+    #updateApiHostProperty(): void {
```
Comment (splincode) [inline projects/addon-doc/components/documentation/directives/documentation-property-connector.directive.ts:191]: too long method, can we decompose it?

File: projects/addon-doc/components/documentation/directives/documentation-property-connector.directive.ts (hunk 2 above)
Comment (MillerSvt) [reply projects/addon-doc/components/documentation/directives/documentation-property-connector.directive.ts:186]: I don't understand why... the tree is created only for this function and is used only in it.

File: projects/addon-doc/components/documentation/directives/documentation-property-connector.directive.ts [hunk 4]
```diff
@@ -188,47 +188,73 @@ export class TuiDocDocumentationPropertyConnectorDirective<T>
         this.locationRef.go(this.urlStateHandler(tree));
     }
 
-    #updateApiHostProperty(): void {
+    private updateApiHostProperty(): void {
         if (this.documentationPropertyPreventUpdateApiHost) {
             return;
         }
 
+        const {documentationPropertyMode: mode, documentationPropertyName: name} = this;
```
Comment (splincode) [inline projects/addon-doc/components/documentation/directives/documentation-property-connector.directive.ts:196]: you can just

```ts
    @Input(`documentationPropertyName`)
    name = '';

    @Input(`documentationPropertyMode`)
    mode: TuiDocumentationPropertyType = null;

    @Input(`documentationPropertyType`)
    type = '';

    @Input(`documentationPropertyValue`)
    value!: T;

    @Input(`documentationPropertyDeprecated`)
    deprecated = false;

    @Input(`documentationPropertyDefaultValue`)
    defaultValue?: T;

    @Input(`documentationPropertyPreventUpdateApiHost`)
    updateApiHost = false;

    @Input(`documentationPropertyValues`)
    values: readonly T[] | null = null;
```

Comment (splincode) [reply projects/addon-doc/components/documentation/directives/documentation-property-connector.directive.ts:196]: sure, sorry

**Comment by MillerSvt** [comment]
Will there be any more comments or can I already write tests?

**Comment by MillerSvt** [comment]
@splincode @waterplea ready to review

File: projects/addon-doc/components/demo/demo.style.less [hunk 5]
```diff
@@ -5,37 +5,44 @@
...
-    border: 1px solid var(--tui-base-03);
-    border-radius: var(--tui-radius-m);
     max-width: 100%;
     min-height: 6rem;
     width: 100%;
     min-width: 10rem;
     background-color: var(--tui-base-01);
     overflow: hidden;
+    position: relative;
+    z-index: 2;
```
Comment (waterplea) [inline projects/addon-doc/components/demo/demo.style.less:44]: Are `position` and `z-index` relative? Doesn't seem like they do anything.

File: projects/addon-doc/components/demo/demo.style.less [hunk 6]
```diff
@@ -177,23 +180,41 @@
...
 .t-form-controls {
     margin-top: 0.75rem;
 }
 
+.t-code-container {
+    background: var(--tui-base-01);
+    border-radius: 0 0 var(--tui-radius-m) var(--tui-radius-m);
+}
+
+.t-code {
```
Comment (waterplea) [inline projects/addon-doc/components/demo/demo.style.less:198]: Doesn't seem like any of these styles do anything. If they are not needed — don't forget to drop the class from the template.

File: projects/addon-doc/components/demo/demo.style.less [hunk 7]
```diff
@@ -177,23 +180,41 @@
...
     @media screen and (min-height: 37.5rem) {
         :host {
-            &._sticky {
+            &._sticky._allow_sticky {
                 position: sticky;
-                top: 4.625rem;
+                top: 0;
                 z-index: 2;
+
+                > .t-wrapper {
```
Comment (waterplea) [inline projects/addon-doc/components/demo/demo.style.less:214]: I don't think this is needed, I have disabled those styles and everything still works as expected.

**Review by waterplea — COMMENTED** [review]
This is a bit overwhelming, it seems like this can be done simpler. Can we have one global `providedIn: "root"` service so we do not need to provide it everywhere? The demo component can just reset it when it's loaded since there's only one of them we need to support at a time, and if we will need more — we can always provide the service.

Also the service itself looks very complicated. What is the purpose of all those streams? Shouldn't it be just `setTemplate` to reset the service to work with a new component and then all the documentation props will call `setProperty` to add/remove/update the bindings?

File: projects/addon-doc/components/demo/demo.template.html [hunk 8]
```diff
@@ -1,89 +1,71 @@
...
+    >
+        Code
+    </button>
+    <tui-hosted-dropdown
+        [content]="dropdownContent"
+        [(open)]="optionsDropdownOpen"
+    >
+        <button
+            tuiTab
+            (tui-tab-activate.stop)="(null)"
```
[the remark this answers is not in the document]
Comment (MillerSvt) [reply projects/addon-doc/components/demo/demo.template.html:19]: fixed

File: projects/addon-doc/components/demo/demo.template.html [hunk 9]
```diff
@@ -1,89 +1,71 @@
...
+        </button>
+    </tui-hosted-dropdown>
+</tui-tabs>
+
+<ng-template #dropdownContent>
+    <tui-data-list
+        role="menu"
+        size="s"
+        class="t-options"
+        [tuiMode]="null"
```
[the remark this answers is not in the document]
Comment (MillerSvt) [reply projects/addon-doc/components/demo/demo.template.html:32]: fixed

File: projects/addon-doc/components/demo/demo.template.html [hunk 10]
```diff
@@ -1,89 +1,71 @@
...
-            <tui-radio-block
-                item="onLight"
-                nativeId="form"
-                size="s"
-                class="tui-group__auto-width-item"
-                [hideRadio]="true"
-                [ngModel]="mode"
-                (ngModelChange)="onModeChange($event)"
+                <div class="t-mode-option">
+                    <tui-radio
```
[the remark this answers is not in the document]
Comment (MillerSvt) [reply projects/addon-doc/components/demo/demo.template.html:41]: ok

File: projects/addon-doc/components/demo/demo.template.html [hunk 11]
```diff
@@ -177,3 +159,16 @@
...
+
+<ng-container *ngIf="activeItemIndex === 1">
+    <tui-elastic-container
+        *ngIf="code$ | async as code"
+        class="t-code-container"
+        [tuiMode]="null"
+    >
+        <tui-doc-code
+            class="t-code"
+            [code]="code!"
```
[the remark this answers is not in the document]
Comment (MillerSvt) [reply projects/addon-doc/components/demo/demo.template.html:171]: fixed

File: projects/addon-doc/components/demo/demo.style.less [hunk 12]
```diff
@@ -177,23 +180,41 @@
...
+
+.t-code {
+    position: relative;
+    z-index: 1;
+    overflow: hidden;
+    height: 100%;
+    --tui-code-action-top: 1.5rem;
+}
+
 @supports (position: sticky) {
```
[the remark this answers is not in the document]
Comment (MillerSvt) [reply projects/addon-doc/components/demo/demo.style.less:206]: fixed

File: projects/addon-doc/components/demo/demo.style.less [hunk 13]
```diff
@@ -177,23 +180,41 @@
     }
 }
 
-.t-group,
 .t-checkbox {
     margin-left: 0.75rem;
 }
 
 .t-button,
+.t-copy-code,
```
[the remark this answers is not in the document]
Comment (MillerSvt) [reply projects/addon-doc/components/demo/demo.style.less:188]: fixed

File: projects/addon-doc/components/demo/demo.component.ts [hunk 14]
```diff
@@ -92,8 +98,30 @@ export class TuiDocDemoComponent implements OnInit {
...
+        private readonly apiHostService: TuiApiHostService | null,
     ) {}
 
+    get code$(): Observable<string> | undefined {
+        return this.apiHostService?.code$;
+    }
+
+    @HostBinding('class._allow_sticky')
+    get allowSticky(): boolean {
+        return this.activeItemIndex === 0;
```
[the remark this answers is not in the document]
Comment (MillerSvt) [reply projects/addon-doc/components/demo/demo.component.ts:112]: fixed

File: projects/addon-doc/components/code/code.style.less [hunk 15]
```diff
@@ -30,8 +30,8 @@
 
 .t-code-actions {
     position: absolute;
-    top: 0.75rem;
-    right: 0.75rem;
+    top: var(--tui-code-action-top, 0.875rem);
```
[the remark this answers is not in the document]
Comment (MillerSvt) [reply projects/addon-doc/components/code/code.style.less:33]: fixed

File: projects/addon-doc/components/demo/demo.template.html [hunk 16]
```diff
@@ -1,89 +1,71 @@
...
+    <button tuiTab>Preview</button>
+    <button
+        *ngIf="code$ | async"
+        tuiTab
+    >
+        Code
+    </button>
+    <tui-hosted-dropdown
+        [content]="dropdownContent"
+        [(open)]="optionsDropdownOpen"
```
[the remark this answers is not in the document]
Comment (MillerSvt) [reply projects/addon-doc/components/demo/demo.template.html:15]: This is not a TuiButton. This is TuiTab.

File: projects/addon-doc/components/demo/demo.style.less (hunk 6 above)
Comment (MillerSvt) [reply projects/addon-doc/components/demo/demo.style.less:198]: This is not need anymore after redesign. Fixed.

File: projects/addon-doc/components/demo/demo.style.less (hunk 5 above)
Comment (MillerSvt) [reply projects/addon-doc/components/demo/demo.style.less:44]: This is not need anymore after redesign. Fixed.

File: projects/addon-doc/components/demo/demo.style.less (hunk 7 above)
Comment (MillerSvt) [reply projects/addon-doc/components/demo/demo.style.less:214]: fixed

**Comment by MillerSvt** [comment]
> Can we have one global providedIn: "root" service so we do not need to provide it everywhere?

Let's try to just add the provider to the tui-doc-page

> What is the purpose of all those streams?

When some component setting the template, component should store it's own index of template, and pass it anytime when template updated. Because template should be update at the same position as before. For avoid it, I create stream, that can store index of template itself. If you have any questions, we can discuss them on meet.

**Pull request closed without merging**